### PR TITLE
feat: quorum certificate validation

### LINKED ISSUE #149: [HotStuff] Add a merkle proof to the QC showing that the committee is valid
The QC should have a merkle proof of the vns in the QC that have signed, and it's root should match the VN_SET_ROOT in the header for that this epoch.

When receiving a QC, if this merkleproof does not match, it should be discarded as invalid and a warning shown.

### PR DESCRIPTION
**Pull request by mrnaveira**
Description
---
* Refactored `ValidatorSignature` into `ValidatorMetadata` to add:
    * Signer public key
    * Merkle proof of the inclusion in the base layer VN merkle root
    * Merkle leaf node index of the signer for the merkle proof
* `VoteMessage` now includes a `ValidatorMetadata` with all the new info.
* `QuorumCertificate` now includes the list of all the `ValidatorMetadata` of the signing committee.
* The Epoch Manager now retrieves the VN merkle root from the base layer and stores it in sqlite (associated with an epoch)
* QC validation:
    * There are no repeated signers
    * The minimum quorum has been reached
    * All merkle proofs for the signers are valid (checked against the merkle root from the base layer)
    * All signers are included in the shard's committee
    * All signatures are valid

There are some improvements that could be made in the `tari_mmr` project to improve this work:
* Allow merkle mountain range serialization, to store in the VN's database (to avoid recalculating it)
* Allow proofs for multiple hashes (to avoid needing multiple proofs)

Motivation and Context
---
HotStuff `VoteMessages` and `QuorumCertificates` must be properly signed and validated.

Closes https://github.com/tari-project/tari-dan/issues/149

How Has This Been Tested?
---
Unit and integration tests pass

### CODE REVIEW

**Review by stringhandler — COMMENTED** [review]
Looking great so far

File: dan_layer/core/src/workers/hotstuff_waiter.rs [hunk 1]
```diff
@@ -364,12 +375,75 @@ where
...
+        // extract all the pairs of signer-signature present in the QC
+        let signer_signatures = Self::extract_signer_signatures_from_qc(qc)?;
+
+        // the QC should not have repeated signers
+        let signers_vec: Vec<&[u8]> = signer_signatures
+            .iter()
+            .map(|s| NodeAddressable::as_bytes(&s.0))
+            .collect();
+        let signers_set: HashSet<&[u8]> = HashSet::from_iter(signers_vec.to_owned());
+        if signers_vec.len() != signers_set.len() {
```
Comment (sdbondi) [inline dan_layer/core/src/workers/hotstuff_waiter.rs]: might work:
```suggestion
        let signers_iter = signer_signatures
            .iter()
            .map(|s| NodeAddressable::as_bytes(&s.0).to_owned())
            ;
        let signers_set: HashSet<&[u8]> = HashSet::from_iter(signers_iter);
        if signature_signers.len() != signers_set.len() {
```

**Review by sdbondi — APPROVED** [review]
LGTM

File: dan_layer/core/src/workers/hotstuff_waiter.rs [hunk 2]
```diff
@@ -653,12 +666,72 @@ where
...
+        let committee = self.epoch_manager.get_committee(qc.epoch(), qc.shard()).await?;
+        if signers_set.len() < committee.consensus_threshold() {
+            return Err(HotStuffError::InvalidQuorumCertificate(
+                "insufficient quorum".to_string(),
+            ));
+        }
+
+        // All signers must be included in the epoch commitee for the shard
+        let commitee_iter = committee.members.iter().map(|m| m.as_bytes());
+        let commitee_set: HashSet<&[u8]> = HashSet::from_iter(commitee_iter);
```
Comment (jorgeantonio21) [inline dan_layer/core/src/workers/hotstuff_waiter.rs]: I think you can pass directly `committee.members.iter().map(|m| m.as_bytes()).collect::<HashSet<_>>()`.

File: dan_layer/core/src/models/hot_stuff_message.rs [hunk 3]
```diff
@@ -154,11 +154,11 @@ impl<TPayload: Payload, TAddr: NodeAddressable> HotStuffMessage<TPayload, TAddr>
         self.message_type() == message_type && view_id == self.view_number()
     }
 
-    pub fn add_partial_sig(&mut self, _signature: ValidatorSignature) {
```
Comment (stringhandler) [inline dan_layer/core/src/models/hot_stuff_message.rs:157]: These two methods can be deleted

**Review by stringhandler — APPROVED** [review]
Looks good. Will have to test it locally, but I think it can go in

**Review by sdbondi — APPROVED** [review]
lightly tested but looked at the code and LGTM

**Pull request merged**